crates/tracing/src/lib.rs

```diff
@@ -0,0 +1,220 @@
+//! Miralis ABI
+//!
+//! While miralis forwards standard SBI calls to the virtualized firmware, miralis do expose its own
+//! ABI for the firmware to interact with.
+//!
+//! The Miralis ABI tries to be compatible with the SBI specification as best as it can.
+//! See: https://github.com/riscv-non-isa/riscv-sbi-doc
+#![no_std]
+
+use core::arch::{asm, global_asm};
+
+use miralis_abi::failure;
+
+pub fn print_statistics(stats: Statistics) {
+    log::info!("{:?}", stats);
+}
+
+#[macro_export]
+macro_rules! measure {
+    ($policy_name:literal, $out:literal, $in:expr, $in2:expr) => {
+        let mut values: [usize; NB_REPEATS] = [0; NB_REPEATS];
+
+        for i in 0..NB_REPEATS {
+            values[i] = $in();
+        }
+
+        let stats = get_statistics(values);
+
+        let average_measure = $in2();
+
+        log::info!("{} cost {} : {}", $policy_name, $out, average_measure);
+
+        print_statistics(stats);
+    };
+}
+
+pub fn enable_mcycle_in_smode() {
+    unsafe {
+        // This allows to read cycle in S-mode - for the payload
+        let mcounteren: u32;
+        asm!("csrr {}, mcounteren", out(reg) mcounteren);
+        asm!("csrw mcounteren, {}", in(reg) mcounteren | 1);
+    }
+}
+
+// —————————————————————————————— Trap Handler —————————————————————————————— //
+
+global_asm!(
+    r#"
+.text
+.align 4
+.global _empty_handler
+_empty_handler:
+    // Skip illegal instruction (pc += 4)
+    csrrw x5, mepc, x5
+    addi x5, x5, 4
+    csrrw x5, mepc, x5
+    // Return back to OS
+    mret
+"#,
+);
```

Comment:
In libraries we don't want to insert global assembly, because it can override existing symbols and cause at best linker errors, at worst arbitrarily bad things.